File: crex.cpp
/* Crop Extractor
 * Copyright (C) 2011 Joel Granados <joel.granados@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */
#include <opencv/cv.h>
#include <opencv/highgui.h>
#include<fstream>
#include "crex.h"

using namespace cv;
using namespace std;

/*{{{ CroppedImage Class */
CroppedImage::CroppedImage ():
  image(Mat::zeros(1,1,CV_8U)),label("") {}

CroppedImage::CroppedImage ( const Mat& image, const string& label ):
  image(image),label(label) {}

bool
CroppedImage::saveTo ( const string& filename )
{
  //FIXME: Does this throw an exception?
  return imwrite(filename,this->image);
}

bool
CroppedImage::adjustDims ( const int width, const int height )
{
  Mat temp = Mat::zeros ( 1, 1, CV_32F );
  Point2f srcpoints[4] = {
    Point2f ( 0, 0 ),
    Point2f ( 0, this->image.size().height ),
    Point2f ( this->image.size().width, 0 ),
    Point2f ( this->image.size().width, this->image.size().height ) };
  Point2f dstpoints[4] = {
    Point2f ( 0, 0 ),
    Point2f ( 0, height ),
    Point2f ( width, 0 ),
    Point2f ( width, height ) };

  Mat aftr = getAffineTransform ( srcpoints, dstpoints );
  warpAffine ( this->image, temp, aftr, Size(width, height) );
  this->image = temp;

  return true;
}

string
CroppedImage::get_label () {return label;}
/*}}} CroppedImage Class*/

/*{{{ VirtualCroppedImageExtractor Class */
VirtualCroppedImageExtractor::VirtualCroppedImageExtractor ():
  imgfile(""),
  annfile(""),
  img(Mat::zeros(1,1,CV_8U)),
  croppedImages(vector<CroppedImage*>()) {}

VirtualCroppedImageExtractor::VirtualCroppedImageExtractor
  ( const string& imgfile, const string& annfile ):
  imgfile(imgfile),
  annfile(annfile),
  img(Mat::zeros(1,1,CV_8U)),
  croppedImages(vector<CroppedImage*>())
{
  img = imread(this->imgfile);
}

bool
VirtualCroppedImageExtractor::extractCroppedImages ()
{
  vector<SimpleAnn> rectangles = getRectangles();
  vector<SimpleAnn>::iterator iter;
  CroppedImage *ci;

  this->croppedImages.clear();
  for ( iter=rectangles.begin() ; iter != rectangles.end(); iter++ )
  {
    ci = new CroppedImage ( Mat(this->img, (*iter).rect), (*iter).label );
    this->croppedImages.push_back( ci );
  }
  return true;
}

vector<CroppedImage*>
VirtualCroppedImageExtractor::getCroppedImages ()
{
  return this->croppedImages;
}
/*}}} VirtualCroppedImageExtractor Class */

/*{{{ ITUAnnotationVer1*/
ITUAnnotationVer1::ITUAnnotationVer1():VirtualCroppedImageExtractor(){}

ITUAnnotationVer1::ITUAnnotationVer1
  ( const string& imgfile, const string& annfile ):
  VirtualCroppedImageExtractor ( imgfile, annfile ){};

vector<SimpleAnn>
ITUAnnotationVer1::getRectangles ()
{
  string line;
  vector<SimpleAnn> rectangles;
  ifstream annfilestream ( this->annfile.data(), ifstream::in );
  int from, to;

  if ( annfilestream.is_open() )
  {
    while ( annfilestream.good() )
    {
      getline(annfilestream, line);

      /* Ignore everything except lines starting with "Bounding"*/
      if ( line.substr(0,8).compare("Bounding") != 0 || line.size() < 75)
        continue;

      // FIXME: Need error checking.
      string label;
      Rect rect;

      /*label*/
      from=line.find(" \"")+2; to=line.find("\" (", from);
      label = line.substr(from, to-from);

      /*xmin*/
      from=to+33; to=line.find(", ", from);
      istringstream(line.substr(from,to-from)) >> rect.x;

      /*ymin*/
      from=to+2; to=line.find(") - (", from);
      istringstream(line.substr(from,to-from)) >> rect.y;

      /*width - xmax*/
      from=to+5; to=line.find(", ", from);
      istringstream(line.substr(from,to-from)) >> rect.width;
      rect.width = rect.width - rect.x;

      /*height - ymax*/
      from=to+2; to =line.find(")", from);
      istringstream(line.substr(from,to-from)) >> rect.height;
      rect.height = rect.height - rect.y;

      SimpleAnn temp;
      temp.rect = rect;
      temp.label = label;
      rectangles.push_back ( temp );
    }
    annfilestream.close();
  }
  return rectangles;
}

/*}}} ITUAnnotationVer1*/

int
main (){
  CroppedImageExtractor cie("file", "file2");
}

File: crex.py
#!/usr/bin/env python
# coding=utf-8
# Crop Extractor
# Copyright (C) 2011 Joel Granados <joel.granados@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
#
import os
import os.path
import pycrex
import logging
import sys

def crex_dir ( from_dir, to_dir=None, ext="png", new_size=None, margin=0 ):
    """Extract annotated images.
    from_dir = Where to look for the images and annotations.
    to_dir = Where to put the crops.  Defaults to from_dir.
    ext = Extension used to save the crops.
    new_size = Size normalization param. (width, height)
    """

    if to_dir==None:
        to_dir = from_dir

    if not os.path.isdir(from_dir):
        raise CREXDirException(from_dir)
    if not os.path.isdir(to_dir):
        raise CREXDirException(to_dir)

    ### CREATE ann_pairs ###
    ann_pairs = [] #[(file, anno), (file, anno)...]
    for entry in os.listdir(from_dir):
        ann_file = os.path.join(from_dir,entry)
        if os.path.isdir( ann_file ) \
           or not ann_file.endswith(".ann"):
            continue

        # here, ann_file ends with ".ann" and is a file.
        img_file = os.path.splitext(ann_file)[0] #img_file contains full path

        if not os.path.exists(img_file) or os.path.isdir(img_file):
            continue #.ann file does not have a pair

        ann_pairs.append( (img_file, ann_file) )

    ### DO THE CREX ###
    npf = NumberedPathFactory(to_dir, ext=ext)
    for ann_pair in ann_pairs:
        (img_file, ann_file) = ann_pair
        crexlog.debug("crexing %s" % ann_file)
        crop_imgs = pycrex.get_cropped_images(img_file, ann_file, margin)
        for crop_img in crop_imgs:
            if not new_size is None and type(new_size) is tuple \
                    and len(new_size) == 2:
                crop_img.adjust_dims( new_size[0], new_size[1] )

            crop_img.save_to( npf.get_output_path(crop_img.get_label()) )

class NumberedPathFactory:
    #top_dir = None
    #label_ref = None

    def __init__(self, top_dir, ext="png" ):
        self.top_dir = top_dir
        self.label_ref = {}
        self.ext = ext

    def get_output_path(self, label):
        label_dir = os.path.join (self.top_dir, label)
        if not label in self.label_ref.keys():#create new label entry.
            if os.path.isdir(label_dir):
                #Search for label offset in label_dir
                clo = -1 #Current Label Offset. -1 in case there are no files
                for entry in os.listdir(label_dir):
                    if os.path.isdir( os.path.join(label_dir,entry) ):
                        continue

                    try:
                        # Assume info is between ^ and first "."
                        tlo = int(entry.split(".")[0]) #Temp Label Offset
                    except:
                        continue
                    if tlo > clo:
                        clo = tlo
                self.label_ref[label] = self.pfn(clo+1)

            else:
                os.mkdir(label_dir)
                self.label_ref[label] =  self.pfn(0)

        retval = os.path.join(label_dir,self.label_ref[label])
        self.label_ref[label] = self.pfn( int(self.label_ref[label]) + 1 )
        return retval+"."+self.ext


    # Pretty Format Number.
    def pfn(self, number, digits=6):
        return ((digits*"0"+"%d")%number)[-digits:]

class CREXException(Exception):
    message = "Unknown error"
    def __init__(self):
        pass
    def __str__(self):
        return ("CREX Error: %s" % self.message)
class CREXDirException(CREXException):
    def __init__(self, dir_name):
        self.message = "Directory %s not found"%dir_name

def initLogger():
    Logger = logging.getLogger("crex")
    Logger.setLevel(logging.DEBUG)
    handler = logging.StreamHandler(sys.stdout)

    #FIXME: Add a file log when we add the config file.
    #handler = logging.FileHandler(config.log.filename)

    formatter = logging.Formatter("%(asctime)s -%(levelname)s - %(message)s")
    handler.setFormatter(formatter)
    Logger.addHandler(handler)

initLogger()
crexlog = logging.getLogger("crex")

